instrumentation/apache-httpclient/apache-httpclient-5.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ApacheHttpClientInstrumentation.java

```diff
@@ -121,242 +124,232 @@ public void transform(TypeTransformer transformer) {
-        @Advice.Local("otelFullRequest") ClassicHttpRequest fullRequest,
-        @Advice.Local("otelContext") Context context,
-        @Advice.Local("otelScope") Scope scope) {
-      Context parentContext = currentContext();
-      fullRequest = new RequestWithHost(host, request);
-      if (!instrumenter().shouldStart(parentContext, fullRequest)) {
-        return;
-      }
+    public static AdviceLocals methodEnter(
+        @Advice.Argument(0) HttpHost host, @Advice.Argument(1) ClassicHttpRequest request) {
 
-      context = instrumenter().start(parentContext, fullRequest);
-      scope = context.makeCurrent();
+      return AdviceLocals.start(new RequestWithHost(host, request));
     }
 
     @Advice.OnMethodExit(onThrowable = Throwable.class, suppress = Throwable.class)
     public static void methodExit(
         @Advice.Return Object result,
         @Advice.Thrown Throwable throwable,
-        @Advice.Local("otelFullRequest") ClassicHttpRequest fullRequest,
-        @Advice.Local("otelContext") Context context,
-        @Advice.Local("otelScope") Scope scope) {
-      if (scope == null) {
-        return;
-      }
+        @Advice.Enter @Nullable AdviceLocals locals) {
 
-      scope.close();
-      ApacheHttpClientHelper.doMethodExit(context, fullRequest, result, throwable);
+      if (locals != null) {
+        locals.end(result, throwable);
+      }
     }
   }
 
   @SuppressWarnings("unused")
   public static class RequestWithHostAndHandlerAdvice {
 
+    @AssignReturned.ToArguments(@ToArgument(value = 2, index = 1))
     @Advice.OnMethodEnter(suppress = Throwable.class)
-    public static void methodEnter(
+    public static Object[] methodEnter(
         @Advice.Argument(0) HttpHost host,
         @Advice.Argument(1) ClassicHttpRequest request,
-        @Advice.Argument(value = 2, readOnly = false) HttpClientResponseHandler<?> handler,
-        @Advice.Local("otelFullRequest") ClassicHttpRequest fullRequest,
-        @Advice.Local("otelContext") Context context,
-        @Advice.Local("otelScope") Scope scope) {
+        @Advice.Argument(2) HttpClientResponseHandler<?> originalHandler) {
 
-      Context parentContext = currentContext();
-      fullRequest = new RequestWithHost(host, request);
-      if (!instrumenter().shouldStart(parentContext, fullRequest)) {
-        return;
-      }
+      HttpClientResponseHandler<?> handler = originalHandler;
+      AdviceLocals locals = AdviceLocals.start(new RequestWithHost(host, request));
 
-      context = instrumenter().start(parentContext, fullRequest);
-      scope = context.makeCurrent();
+      if (locals == null) {
+        return new Object[] {null, handler};
+      }
 
       // Wrap the handler so we capture the status code
       if (handler != null) {
         handler =
             new WrappingStatusSettingResponseHandler<>(
-                context, parentContext, fullRequest, handler);
+                locals.context, locals.parentContext, locals.request, handler);
```

Comment:
Imo it would be cleaner to add a `wrapResponseHandler(handler)` method to `AdviceLocals`. This allows you to make the `AdviceLocals` fields private and remove direct accesses from the advices to them.